Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp

```diff
@@ -24,4 +28,219 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
+
+    // If the candidate high boundary is less than the low boundary, then the high boundary is the low boundary.
+    // Otherwise, if the candidate high boundary is greater than the maximum value, then the high boundary is the maximum value.
+    // Otherwise, the high boundary is the candidate high boundary.
+    return clamp(candidate_high, min(), max());
+}
+
+WebIDL::ExceptionOr<void> HTMLMeterElement::set_high(double value)
+{
+    TRY(set_attribute(HTML::AttributeNames::high, MUST(String::number(value))));
+    update_meter_value_element();
+    document().invalidate_layout();
+    return {};
+}
+
+// https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-optimum
+double HTMLMeterElement::optimum() const
+{
+    // If the optimum attribute is specified and a value could be parsed out of it, then the candidate optimum point is that value. Otherwise, the candidate optimum point is the midpoint between the minimum value and the maximum value.
+    double candidate_optimum = (max() + min()) / 2;
+    auto maybe_optimum_string = get_attribute(HTML::AttributeNames::optimum);
+    if (maybe_optimum_string.has_value()) {
+        auto maybe_optimum = parse_floating_point_number(maybe_optimum_string.value());
+        if (maybe_optimum.has_value())
+            candidate_optimum = maybe_optimum.value();
+    }
+
+    // If the candidate optimum point is less than the minimum value, then the optimum point is the minimum value.
+    // Otherwise, if the candidate optimum point is greater than the maximum value, then the optimum point is the maximum value.
+    // Otherwise, the optimum point is the candidate optimum point.
+    return clamp(candidate_optimum, min(), max());
+}
+
+WebIDL::ExceptionOr<void> HTMLMeterElement::set_optimum(double value)
+{
+    TRY(set_attribute(HTML::AttributeNames::optimum, MUST(String::number(value))));
+    update_meter_value_element();
+    document().invalidate_layout();
+    return {};
+}
+
+void HTMLMeterElement::inserted()
+{
+    create_shadow_tree_if_needed();
+}
+
+void HTMLMeterElement::create_shadow_tree_if_needed()
+{
+    if (shadow_root_internal())
+        return;
+
+    auto shadow_root = heap().allocate<DOM::ShadowRoot>(realm(), document(), *this, Bindings::ShadowRootMode::Closed);
+    set_shadow_root(shadow_root);
+
+    auto meter_bar_element = heap().allocate<MeterBarElement>(realm(), document());
+    MUST(shadow_root->append_child(*meter_bar_element));
+
+    m_meter_value_element = heap().allocate<MeterValueElement>(realm(), document());
+    MUST(meter_bar_element->append_child(*m_meter_value_element));
+    update_meter_value_element();
+}
+
+void HTMLMeterElement::update_meter_value_element()
+{
+    if (!m_meter_value_element)
+        return;
+
+    // UA requirements for regions of the gauge:
+    double m_value = value();
+    double m_min = min();
+    double m_max = max();
+    double m_low = low();
+    double m_high = high();
+    double m_optimum = optimum();
```

Comment:
Naming these `m_foo` when they're not member variables is a bit odd. I get that's it's hard to find a good name though. Could always do `double foo = this->foo();` if nothing else.